external-crates/move/move-compiler/src/hlir/detect_dead_code.rs

```diff
@@ -339,120 +380,86 @@ fn tail_block(context: &mut Context, seq: &T::Sequence) -> Option<ControlFlow> {
 fn value(context: &mut Context, e: &T::Exp) -> Option<ControlFlow> {
     use T::UnannotatedExp_ as E;
 
-    if is_binop(e) {
-        return process_binops(context, e);
-    }
-
     let T::Exp {
         ty,
         exp: sp!(eloc, e_),
     } = e;
 
+    macro_rules! value_report {
+        ($nested_value:expr) => {{
+            if let Some(control_flow) = value(context, $nested_value) {
+                context.report_value_error(control_flow);
+                already_reported(*eloc)
+            } else {
+                None
+            }
+        }};
+    }
+
     match e_ {
         // -----------------------------------------------------------------------------------------
         // control flow statements
         // -----------------------------------------------------------------------------------------
         E::IfElse(test, conseq, alt) => {
             if let Some(test_control_flow) = value(context, test) {
                 context.report_value_error(test_control_flow);
+                return already_reported(*eloc);
             };
-            let conseq_flow = value(context, conseq);
-            let alt_flow = value(context, alt);
-            match (conseq_flow, alt_flow) {
-                _ if matches!(ty, sp!(_, N::Type_::Unit)) => None,
-                (Some(cflow), Some(aflow)) if cflow.value == aflow.value => {
-                    context.report_value_error(sp(*eloc, cflow.value));
-                    None
-                }
-                (Some(cflow), Some(aflow)) => {
-                    context.report_value_error(cflow);
-                    context.report_value_error(aflow);
-                    None
+            if !matches!(ty, sp!(_, N::Type_::Unit)) {
+                if let (Some(cflow), Some(aflow)) = (value(context, conseq), value(context, alt)) {
```

Comment:
This is for rule 4? Why is it just for unit?

Reply:
This is an artifact left over from when it was weaved into the HLIR lowering itself. Since the type now is always `_` if it's divergent, we can't even trigger the exclusion case. The goal was to allow things like 

```
let () = (if (cond) { break } else { continue } : ());
```

This currently won't work because we report errors before returning, though. We could allow this by suppressing warnings for situations where we expect `()` as a value, but I also think producing  warning when people write code like this is just fine.